File: backend/app/schemas/shared_kitchen.py
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Literal
import uuid

from pydantic import Field

from app.schemas import BaseSchema


class CompanyKitchenConfigureRequest(BaseSchema):
    branch_id: uuid.UUID
    raw_location_id: uuid.UUID
    name: str = Field(min_length=1, max_length=255)
    timezone: str = Field(default="Asia/Bangkok", min_length=1, max_length=50)


class CompanyIngredientCreateRequest(BaseSchema):
    canonical_product_id: uuid.UUID
    code: str = Field(min_length=1, max_length=80)
    name: str = Field(min_length=1, max_length=255)
    base_unit_code: str = Field(min_length=1, max_length=30)
    unit_dimension: Literal["mass", "volume", "count"]


class CompanyIngredientAliasCreateRequest(BaseSchema):
    ingredient_id: uuid.UUID
    brand_id: uuid.UUID
    source_product_id: uuid.UUID
    source_unit_code: str = Field(min_length=1, max_length=30)
    conversion_factor: Decimal = Field(gt=0)
    supplier_sku: str | None = Field(default=None, max_length=100)


class CompanyIngredientReceiptRequest(BaseSchema):
    ingredient_id: uuid.UUID
    lot_code: str = Field(min_length=1, max_length=100)
    qty: Decimal = Field(gt=0)
    unit_cost: Decimal = Field(ge=0)
    expires_on: date | None = None
    idempotency_key: str = Field(min_length=8, max_length=120)
    reference_type: str = Field(default="goods_receipt", min_length=1, max_length=50)
    reference_id: str = Field(min_length=1, max_length=120)
    note: str | None = None


class CompanyProductionDemandCreateRequest(BaseSchema):
    brand_id: uuid.UUID
    branch_id: uuid.UUID
    output_product_id: uuid.UUID
    needed_on: date
    requested_qty: Decimal = Field(gt=0)
    unit_code: str = Field(min_length=1, max_length=30)
    source_type: str = Field(min_length=1, max_length=50)
    source_id: str = Field(min_length=1, max_length=120)
    idempotency_key: str = Field(min_length=8, max_length=120)
    note: str | None = None


class CompanyProductionOrderCreateRequest(BaseSchema):
    brand_id: uuid.UUID
    output_product_id: uuid.UUID
    planned_date: date
    planned_qty: Decimal = Field(gt=0)
    idempotency_key: str = Field(min_length=8, max_length=120)
    demand_id: uuid.UUID | None = None
    recipe_id: uuid.UUID | None = None
    note: str | None = None


class CompanyProductionActualInput(BaseSchema):
    input_id: uuid.UUID
    actual_qty: Decimal = Field(ge=0)


class CompanyProductionCompleteRequest(BaseSchema):
    completion_key: str = Field(min_length=8, max_length=120)
    actual_output_qty: Decimal = Field(gt=0)
    waste_qty: Decimal = Field(default=Decimal("0"), ge=0)
    inputs: list[CompanyProductionActualInput] = Field(default_factory=list)
    note: str | None = None


class CompanyProductionReverseRequest(BaseSchema):
    reversal_key: str = Field(min_length=8, max_length=120)
    reason: str = Field(min_length=3, max_length=500)

File: frontend/src/types/sharedKitchen.ts
import type { SupplyChainRelease } from "@/types/supplyChainRelease";

export type CompanyKitchen = {
  id: string;
  name: string;
  branch_id: string;
  branch_name: string;
  raw_location_id: string;
  raw_location_name: string;
  timezone: string;
  costing_method: "fifo";
  allow_negative_stock: false;
  is_active: boolean;
};

export type CompanyIngredient = {
  id: string;
  code: string;
  name: string;
  canonical_product_id: string;
  base_unit_code: string;
  unit_dimension: "mass" | "volume" | "count";
  qty_on_hand: number | string;
  is_active: boolean;
};

export type CompanyIngredientAlias = {
  id: string;
  ingredient_id: string;
  ingredient_name: string;
  brand_id: string;
  brand_name: string;
  source_product_id: string;
  source_product_name: string;
  source_unit_code: string;
  conversion_factor: number | string;
  supplier_sku: string | null;
};

export type CompanyProductionDemand = {
  id: string;
  brand_id: string;
  brand_name: string;
  branch_id: string;
  branch_name: string;
  output_product_id: string;
  output_product_name: string;
  needed_on: string;
  requested_qty: number | string;
  unit_code: string;
  status: "submitted" | "converted" | "cancelled";
  source_type: string;
  source_id: string;
  note: string | null;
};

export type CompanyProductionInput = {
  id: string;
  ingredient_id: string;
  ingredient_name: string;
  planned_qty: number | string;
  actual_qty: number | string | null;
  base_unit_code: string;
  actual_cost: number | string;
};

export type CompanyProductionOrder = {
  id: string;
  order_number: string;
  brand_id: string;
  brand_name: string;
  output_product_id: string;
  output_product_name: string;
  planned_date: string;
  status: "planned" | "in_progress" | "completed" | "reversed" | "cancelled";
  planned_qty: number | string;
  actual_output_qty: number | string | null;
  waste_qty: number | string;
  output_unit_code: string;
  total_input_cost: number | string;
  output_cost_per_unit: number | string;
  demand_id: string | null;
  recipe_id: string;
  note: string | null;
  inputs: CompanyProductionInput[];
};

export type CompanyKitchenDashboard = {
  kitchen: CompanyKitchen | null;
  ingredients: CompanyIngredient[];
  aliases: CompanyIngredientAlias[];
  demands: CompanyProductionDemand[];
  orders: CompanyProductionOrder[];
  write_enabled: boolean;
  release: SupplyChainRelease;
  setup_options: {
    branches: Array<{ id: string; name: string; code: string }>;
    locations: Array<{ id: string; branch_id: string; name: string; code: string }>;
    brands: Array<{ id: string; name: string; business_type: string }>;
    brand_branches: Array<{ brand_id: string; branch_id: string }>;
    products: Array<{
      id: string;
      name: string;
      sku: string;
      brand_id: string | null;
      inventory_role: "central_raw" | "central_ready";
      unit_code: string | null;
    }>;
  };
};

export type CompanyKitchenReport = {
  date_from: string;
  date_to: string;
  production_by_brand: Array<{
    brand_id: string;
    brand_name: string;
    order_count: number;
    completed_count: number;
    output_qty: number | string;
    waste_qty: number | string;
    input_cost: number | string;
  }>;
  ingredient_usage: Array<{
    brand_id: string;
    brand_name: string;
    ingredient_id: string;
    ingredient_name: string;
    consumed_qty: number | string;
    reversed_qty: number | string;
    net_cost: number | string;
    unit_code: string;
  }>;
  orders: CompanyProductionOrder[];
};
